File: python/creaturebattle.py
import hashlib as H
import sys

salt = '.NaCl.$^d43lwz;)3s.optimize.this'

def mangle(monsters):
    cleaned = [s.lower().strip() for s in monsters]
    return [''.join([s, rs, salt]) for (s, rs) in zip(cleaned, reversed(cleaned))]

def main(monsters):
    if monsters[0] == monsters[1]:
        print 'The %s refuses to fight itself!'%(monsters[0],)
    else:
        mangled_monsters = [H.sha1(m).digest() for m in mangle(monsters)]
        battle = dict(zip(mangled_monsters, monsters))
        print 'The %s has won the battle!'%(battle[sorted(battle.keys())[0]],)

if __name__ == '__main__':
    main(sys.argv[1:3])

File: matlab/creaturebattle.m
function creaturebattle(home, away)
    if strcmp(home, away)
        disp(['The ' home ' refuses to fight itself!']);
    else
        salt = '.NaCl.$^d43lwz;)3s.optimize.this';
        mangled = mangle({home, away}, salt);
        mdhome = java.security.MessageDigest.getInstance('SHA1');
        mdaway = java.security.MessageDigest.getInstance('SHA1');
        mdhome.update(mangled(1).getBytes());
        mdaway.update(mangled(2).getBytes());
        battle = sortrows({[mod(double(mdhome.digest),255)' home]; [mod(double(mdaway.digest),255)' away]});
        disp(['The ' battle{1}(21:end) ' has won the battle!']);
    end
end

function mangled = mangle(monsters, salt)
    mangled = cellfun(@(x, y) java.lang.String([x y salt]), ...
        monsters, monsters(end:-1:1), 'UniformOutput', false);
end